[BE][DTensor] validate the mesh argument in DeviceMesh construction

XilunWu opened the pull request:
## What's in this PR
DeviceMesh's __init__ function now requires all calling ranks to pass the same `mesh` argument.

## Why
We want to enforce SPMD style of programs using DTensor. Before this PR, 2-D Parallel API (e.g. _create_1d_device_mesh) defines different DeviceMesh on different ranks. After this PR, it defines each sub-meshes and simply perform communications on the one that it is associated with.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99094

Differential Revision: [D45165511](https://our.internmc.facebook.com/intern/diff/D45165511)

XilunWu committed:
[BE][DTensor] validate the mesh argument in DeviceMesh construction

[ghstack-poisoned]

XilunWu commented:
TODO: need to change existing TP tests

XilunWu committed:
Update on "[BE][DTensor] validate the mesh argument in DeviceMesh construction"




[ghstack-poisoned]

XilunWu committed:
Update on "[BE][DTensor] validate the mesh argument in DeviceMesh construction"




[ghstack-poisoned]

XilunWu committed:
Update on "[BE][DTensor] validate the mesh argument in DeviceMesh construction"




[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/test_device_mesh.py:
```diff
@@ -94,6 +94,17 @@ def test_lazy_init_device_mesh(self):
+    @with_comms
+    def test_validate_device_mesh(self):
+        mesh = torch.arange(self.world_size).reshape(2, -1)
+        mesh_subgp_1 = mesh[0]
```
```suggestion
        mesh_subpg_1 = mesh[0]
```

wanchaol commented on test/distributed/_tensor/test_device_mesh.py:
```diff
@@ -94,6 +94,17 @@ def test_lazy_init_device_mesh(self):
+    def test_validate_device_mesh(self):
+        mesh = torch.arange(self.world_size).reshape(2, -1)
+        mesh_subgp_1 = mesh[0]
+        mesh_subgp_2 = mesh[1]
```
```suggestion
        mesh_subpg_2 = mesh[1]
```

wanchaol reviewed (approved):
Looks good! Thanks for working on it!

Left a small nit. Could you import this to internal too to make sure nothing breaks

XilunWu committed:
Update on "[BE][DTensor] validate the mesh argument in DeviceMesh construction"

## What's in this PR
DeviceMesh's __init__ function now requires all calling ranks to pass the same `mesh` argument.

## Why
We want to enforce SPMD style of programs using DTensor. Before this PR, 2-D Parallel API (e.g. _create_1d_device_mesh) defines different DeviceMesh on different ranks. After this PR, it defines each sub-meshes and simply perform communications on the one that it is associated with.




[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45165511).

XilunWu commented:
@pytorchmergebot merge
